homeassistant/components/todoist/calendar.py

```diff
@@ -562,45 +612,45 @@ def select_best_task(project_tasks):
                 continue
 
             if proposed_event[PRIORITY] == event[PRIORITY] and (
-                proposed_event[END] < event[END]
+                event[END] is not None and proposed_event[END] < event[END]
             ):
                 event = proposed_event
                 continue
-
         return event
 
-    async def async_get_events(self, hass, start_date, end_date):
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
         """Get all tasks in a specific time frame."""
         if self._id is None:
+            tasks = await self._api.get_tasks()
             project_task_data = [
                 task
-                for task in self._api.state[TASKS]
+                for task in tasks
                 if not self._project_id_whitelist
-                or task[PROJECT_ID] in self._project_id_whitelist
+                or task.project_id in self._project_id_whitelist
             ]
         else:
-            project_data = await hass.async_add_executor_job(
-                self._api.projects.get_data, self._id
-            )
-            project_task_data = project_data[TASKS]
+            project_task_data = await self._api.get_tasks(project_id=self._id)
 
         events = []
         for task in project_task_data:
-            if task["due"] is None:
+            if task.due is None:
                 continue
             # @NOTE: _parse_due_date always returns the date in UTC time.
             due_date: datetime | None = _parse_due_date(
-                task["due"], self._api.state["user"]["tz_info"]["hours"]
+                task.due, self._utc_offset_hours
             )
             if not due_date:
                 continue
-            midnight = dt.as_utc(
-                dt.parse_datetime(
-                    due_date.strftime("%Y-%m-%d")
-                    + "T00:00:00"
-                    + self._api.state["user"]["tz_info"]["gmt_string"]
-                )
+            gmt_string = f"{'-' if self._utc_offset_hours < 0 else '+'}{abs(self._utc_offset_hours):02}:00"
+            local_midnight = dt.parse_datetime(
+                f'{due_date.strftime("%Y-%m-%d")}T00:00:00{gmt_string}'
             )
+            if local_midnight is not None:
+                midnight = dt.as_utc(local_midnight)
+            else:
+                midnight = due_date.replace(hour=0, minute=0, second=0, microsecond=0)
```

Comment:
Small change here since mypy correctly pointed out `dt.parse_datetime` could return None and `dt.as_utc` only allows a datetime instance.